### PR TITLE
posix: pthread: implement pthread_cleanup_push() / pop()

### LINKED ISSUE #59940: posix: implement `pthread_cleanup_pop()`
There are currently 20 missing POSIX functions with trivial implementations, `pthread_cleanup_pop()` is one of them.

> pthread_cleanup_pop, pthread_cleanup_push - establish cancellation handlers
> 
> The pthread_cleanup_pop() function shall remove the routine at the top of the calling thread's cancellation cleanup stack and optionally invoke it (if execute is non-zero).
> 
> The pthread_cleanup_push() function shall push the specified cancellation cleanup handler routine onto the calling thread's cancellation cleanup stack. The cancellation cleanup handler shall be popped from the cancellation cleanup stack and invoked with the argument arg when:
> 
> The thread exits (that is, calls [pthread_exit()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_exit.html)).
> 
> The thread acts upon a cancellation request.
> 
> The thread calls pthread_cleanup_pop() with a non-zero execute argument.
> 
> It is unspecified whether pthread_cleanup_push() and pthread_cleanup_pop() are macros or functions. If a macro definition is suppressed in order to access an actual function, or a program defines an external identifier with any of these names, the behavior is undefined. The application shall ensure that they appear as statements, and in pairs within the same lexical scope (that is, the pthread_cleanup_push() macro may be thought to expand to a token list whose first token is '{' with pthread_cleanup_pop() expanding to a token list whose last token is the corresponding '}' ).
> 
> The effect of calling [longjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/longjmp.html) or [siglongjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/siglongjmp.html) is undefined if there have been any calls to pthread_cleanup_push() or pthread_cleanup_pop() made without the matching call since the jump buffer was filled. The effect of calling [longjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/longjmp.html) or [siglongjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/siglongjmp.html) from inside a cancellation cleanup handler is also undefined unless the jump buffer was also filled in the cancellation cleanup handler.
> 
> The effect of the use of return, break, continue, and goto to prematurely leave a code block described by a pair of pthread_cleanup_push() and pthread_cleanup_pop() functions calls is undefined.

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cleanup_pop.html

See RFC #51211 for more info.

### LINKED ISSUE #59941: posix: implement `pthread_cleanup_push()`
There are currently 20 missing POSIX functions with trivial implementations, `pthread_cleanup_push()` is one of them.

> pthread_cleanup_pop, pthread_cleanup_push - establish cancellation handlers
> 
> The pthread_cleanup_pop() function shall remove the routine at the top of the calling thread's cancellation cleanup stack and optionally invoke it (if execute is non-zero).
> 
> The pthread_cleanup_push() function shall push the specified cancellation cleanup handler routine onto the calling thread's cancellation cleanup stack. The cancellation cleanup handler shall be popped from the cancellation cleanup stack and invoked with the argument arg when:
> 
> The thread exits (that is, calls [pthread_exit()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_exit.html)).
> 
> The thread acts upon a cancellation request.
> 
> The thread calls pthread_cleanup_pop() with a non-zero execute argument.
> 
> It is unspecified whether pthread_cleanup_push() and pthread_cleanup_pop() are macros or functions. If a macro definition is suppressed in order to access an actual function, or a program defines an external identifier with any of these names, the behavior is undefined. The application shall ensure that they appear as statements, and in pairs within the same lexical scope (that is, the pthread_cleanup_push() macro may be thought to expand to a token list whose first token is '{' with pthread_cleanup_pop() expanding to a token list whose last token is the corresponding '}' ).
> 
> The effect of calling [longjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/longjmp.html) or [siglongjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/siglongjmp.html) is undefined if there have been any calls to pthread_cleanup_push() or pthread_cleanup_pop() made without the matching call since the jump buffer was filled. The effect of calling [longjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/longjmp.html) or [siglongjmp()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/siglongjmp.html) from inside a cancellation cleanup handler is also undefined unless the jump buffer was also filled in the cancellation cleanup handler.
> 
> The effect of the use of return, break, continue, and goto to prematurely leave a code block described by a pair of pthread_cleanup_push() and pthread_cleanup_pop() functions calls is undefined.

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cleanup_push.html

See RFC #51211 for more info.

### PR DESCRIPTION
**Pull request by cfriedt**
`pthread_cleanup_push()` and `pthread_cleanup_pop()` are required by the POSIX_THREADS_BASE Option Group as detailed in Section E.1 of IEEE-1003.1-2017.

The POSIX_THREADS_BASE Option Group is required for PSE51, PSE52, PSE53, and PSE54 conformance, and is otherwise mandatory for any POSIX conforming system as per Section A.2.1.3 of IEEE-1003-1.2017.

Fixes #59940
Fixes #59941

See also https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cleanup_pop.html

### CODE REVIEW

**Comment by cfriedt** [comment]
Cannot reproduce build errors locally. Will rebase.

**Comment by keith-packard** [comment]
> Cannot reproduce build errors locally. Will rebase.

I can.I'm seeing an allocation failure in `test_thrd_reuse`  when `i == 27` when  `pthread_setspecific(__pthread_cleanup_key, &cleanup_head);` is called. Is that allocation ever freed? Or is there some other memory leak?

**Comment by cfriedt** [comment]
> > Cannot reproduce build errors locally. Will rebase.
> 
> I can. I'm seeing an allocation failure in `test_thrd_reuse` when `i == 27` when `pthread_setspecific(__pthread_cleanup_key, &cleanup_head);` is called. It doesn't look like that allocation is ever freed.

It's not ever freed. It doesn't actually count as one of the `CONFIG_MAX_PTHREAD_KEY_COUNT` keys that are user-allocatable.

I'm using a relatively old SDK version 0.16.0. Still not seeing failures (on qemu_x86 and qemu_riscv64).

**Comment by keith-packard** [comment]
> > > Cannot reproduce build errors locally. Will rebase.
> > 
> > 
> > I can. I'm seeing an allocation failure in `test_thrd_reuse` when `i == 27` when `pthread_setspecific(__pthread_cleanup_key, &cleanup_head);` is called. It doesn't look like that allocation is ever freed.
> 
> It's not ever freed. It doesn't actually count as one of the `CONFIG_MAX_PTHREAD_KEY_COUNT` keys that are user-allocatable.

No, the problem is that there is per-thread data allocated in this call and that never gets freed, so you end up running out of the kernel pool after you allocate 1024 bytes (by default) of these things.

> I'm using a relatively old SDK version 0.16.0. Still not seeing failures (on qemu_x86 and qemu_riscv64).

I'm using 0.16.4 and getting failures. Might be good to give that a whirl?

**Comment by cfriedt** [comment]
> No, the problem is that there is per-thread data allocated in this call and that never gets freed, so you end up running out of the kernel pool after you allocate 1024 bytes (by default) of these things.

Ah - I see it now. Yeah, that's kind of an annoyance of using k_malloc for this sort of thing. I know some other libcs allocate pthread keys in a dedicated soft-TLS area.

I wonder if there is a way to implement pthread_cleanup_push() / pop() without using `pthread_setspecific()` - e.g. just on the stack 🤔 

Let me give this one another whirl.

**Comment by cfriedt** [comment]
@keith-packard - should be ready now.

File: include/zephyr/posix/pthread.h [hunk 1]
```diff
@@ -480,6 +480,18 @@ int pthread_atfork(void (*prepare)(void), void (*parent)(void), void (*child)(vo
 int pthread_getconcurrency(void);
 int pthread_setconcurrency(int new_level);
 
+void __z_pthread_cleanup_push(void *cleanup[3], void (*routine)(void *arg), void *arg);
+void __z_pthread_cleanup_pop(int execute);
+
+#define pthread_cleanup_push(_rtn, _arg)                                                           \
+	do /* enforce '{'-like behaviour */ {                                                      \
+		void *_z_pthread_cleanup[3];                                                       \
```
Comment (keith-packard) [inline include/zephyr/posix/pthread.h:488]: Will this cause compiler warnings if you nest these?

**Review by keith-packard — COMMENTED** [review]
This is looking nice and clean now.

File: include/zephyr/posix/pthread.h (hunk 1 above)
Comment (cfriedt) [reply include/zephyr/posix/pthread.h:488]: I don't believe so - there is a test and I don't see any warnings when building it.

**Review by keith-packard — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**